packages/itwin/tree-widget/src/components/trees/common/UsePerformanceReporting.tsx

```diff
@@ -0,0 +1,44 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { useEffect, useRef } from "react";
+
+import type { IModelConnection } from "@itwin/core-frontend";
+
+/** @internal */
+export interface UsePerformanceReportingProps {
+  treeIdentifier: string;
+  iModel: IModelConnection;
+  onPerformanceMeasured?: (featureId: string, duration: number) => void;
+}
+
+/** @internal */
+export interface UsePerformanceReportingResult {
+  onNodeLoaded?: (props: { node: string; duration: number }) => void;
+}
+
+/**
+ * Enables performance reporting for a tree component.
+ * @internal
+ */
+export function usePerformanceReporting(props: UsePerformanceReportingProps): UsePerformanceReportingResult {
+  const { treeIdentifier, iModel, onPerformanceMeasured } = props;
+  const firstLoadRef = useRef(true);
+
+  useEffect(() => {
+    firstLoadRef.current = true;
+  }, [iModel]);
+
+  const onNodeLoaded = ({ node, duration }: { node: string; duration: number }) => {
+    if (firstLoadRef.current && node === "root") {
```

Comment:
I don't think we should log root nodes load as `-hierarchy-level-load`. In my opinion loading root nodes is `initial-load`

Reply:
Switched to always reporting root nodes as `initial-load`.